arangod/IResearch/IResearchAnalyzerFeature.cpp

```diff
@@ -293,6 +296,139 @@ arangodb::SystemDatabaseFeature::ptr getSystemDatabase() {
+      { "text_pt", "{ \"locale\": \"pt.UTF-8\", \"ignored_words\": [ ] }" }, // empty stop word list
+      { "text_ru", "{ \"locale\": \"ru.UTF-8\", \"ignored_words\": [ ] }" }, // empty stop word list
+      { "text_sv", "{ \"locale\": \"sv.UTF-8\", \"ignored_words\": [ ] }" }, // empty stop word list
+      { "text_zh", "{ \"locale\": \"zh.UTF-8\", \"ignored_words\": [ ] }" }, // empty stop word list
+    };
+    static const irs::flags legacyAnalyzerFeatures = { // add norms + frequency/position for by_phrase
+      irs::frequency::type(), // frequency feature
+      irs::norm::type(), // norm feature
+      irs::position::type(), // position feature
+    };
+    static const irs::string_ref legacyAnalyzerType("text");
+    bool success = true;
+
+    // reguster each legacy static analyzer with the current vocbase
+    for (auto& entry: legacyAnalzyers) {
+      auto name = // prefix name with vocbase
+        std::string(vocbase.name()).append(2, ANALYZER_PREFIX_DELIM).append(entry.first);
+      auto& type = legacyAnalyzerType;
+      auto& properties = entry.second;
+      auto result = // result
+        analyzers->emplace(name, type, properties, legacyAnalyzerFeatures);
+
+      if (!result.first) {
+        LOG_TOPIC(WARN, arangodb::iresearch::TOPIC)
+          << "failure while registering a legacy static analyzer '" << name << "' with vocbase '" << vocbase.name() << "'";
+
+        success = false;
+      }
+    }
+
+    return success;
+  }
+}
+
+void registerUpgradeTasks() {
+  auto* upgrade = arangodb::application_features::ApplicationServer::lookupFeature< // find feature
+    arangodb::UpgradeFeature // feature type
+  >("Upgrade");
+
+  if (!upgrade) {
+    return; // nothing to register with (OK if no tasks actually need to be applied)
+  }
+
+  // register legacy static analyzers with each vocbase found in DatabaseFeature
+  // required for backward compatibility for e.g. TOKENS(...) function
+  // NOTE: db-servers do not have a dedicated collection for storing analyzers,
+  //       instead they get their cache populated from coordinators
+  {
+    arangodb::methods::Upgrade::Task task;
+    task.name = "IResearhAnalyzer legacy analyzers";
+    task.description = // description
+      "register legacy static analyzers with each vocbase found in DatabaseFeature";
+    task.systemFlag = arangodb::methods::Upgrade::Flags::DATABASE_ALL;
+    task.clusterFlags = // flags
+      arangodb::methods::Upgrade::Flags::CLUSTER_COORDINATOR_GLOBAL // any 1 single coordinator
+      | arangodb::methods::Upgrade::Flags::CLUSTER_NONE // local server
+      ;
+    task.databaseFlags = arangodb::methods::Upgrade::Flags::DATABASE_UPGRADE;
+    task.action = &iresearchAnalyzerLegacyAnalyzers;
+    upgrade->addTask(std::move(task));
+
+    // FIXME TODO find out why CLUSTER_COORDINATOR_GLOBAL will only work with DATABASE_INIT (hardcoded in Upgrade::clusterBootstrap(...))
```

Comment:
we would need to fix the upgrade detection code for the coordinator case. Since this was not necessary until now, it was not implemented

Reply:
I have no objections to the upgrade detection code being fixed, then the workarounds could be removed both here and in IResearchFeature.
Would you be able to do this change as you know the Upgrade code much better than I?